### PR TITLE
ci: add `fix_quotation` in `linter_patch` to replace `check-quotes.yml`

### LINKED ISSUE #6748: 将 check-quotes 合并到 linter-patch 中
related #4702

个人认为半角语境下编辑时应该不会有人写『“”』，所以『“”』可以直接替换成『「」』。目前的 linter-patch.py 可以完美实现这个引号的替换（在 post process 中直接替换即可），这样就可以去掉 check-quotes.yml，方便用户编辑。

### PR DESCRIPTION
**Pull request by Tiphereth-A**
- [x] 我已认真阅读贡献指南 (contributing guidelines) 和社区公约 (code of conduct)，并遵循了如何参与页及格式手册页的相应规范。

resolve #6748

测试代码和注释采用 AIGC

### CODE REVIEW

**Comment by c-forrest** [comment]
单引号的替换有些没有必要了，反而增大了编辑难度。可以不自动替换单引号吗？

PS：原来的 `check-quotes.yml` 也没有处理单引号的情形。

**Comment by Tiphereth-A** [comment]
> 单引号的替换有些没有必要了，反而增大了编辑难度。可以不自动替换单引号吗？

我觉得这个应该统一处理。要不单引号的改成该行有几个左单引号就换几次 🤔

File: scripts/linter/postprocess.py [hunk 1]
```diff
@@ -40,3 +40,51 @@ def fix_full_stop(md_content: str, **kwargs):
...
+    # Track changes for logging
+    total_changes = 0
+
+    for i, line in enumerate(lines):
+        if line_origins[i] == -1:
+            # Skip placeholder lines (from skip blocks)
+            continue
+
+        modified_line = line.replace('‘', '『')
+        modified_line = modified_line.replace('’', '』')
```
Comment (CoelacanthusHex) [inline scripts/linter/postprocess.py]: For right single quote, it should only replace when not surround by latin characters. if it matched `/(?<=[a-zA-Z])'(?=[a-zA-Z])/`, it should be a abbreviation.

**Comment by Tiphereth-A** [comment]
> 目前的格式手册似乎未涉及引用英文原文时表示缩写的右单引号应该怎么处理

我觉得在这里先保持原样吧，如果有需要再新开 PR 🤔

**Comment by c-forrest** [comment]
如果 wiki 里没有有中文用单引号的例子，单引号只用于英文环境，而且还可能用于撇号（apostrophe ），那么对单引号的任何处理都意味着额外增加了维护成本，而非减少。我觉得这里没有必要面面俱到，毕竟我们也不是出版商。

**Comment by Tiphereth-A** [comment]
> 如果 wiki 里没有有中文用单引号的例子，单引号只用于英文环境，而且还可能用于撇号（apostrophe ），那么对单引号的任何处理都意味着额外增加了维护成本，而非减少。我觉得这里没有必要面面俱到，毕竟我们也不是出版商。

如果这边认可 `“……”` 是不好的，那么 `‘……’` 也应被认为是不好的，所以如果要处理 `“……”` 那也要处理 `‘……’`。至于维护成本上我觉得没有什么区别，这两种情况的代码是基本一致的（如果不处理 apostrophe 那 `inner_replace` 就基本是把相同的代码重复了两遍而已）。

**Comment by c-forrest** [comment]
> > 如果 wiki 里没有有中文用单引号的例子，单引号只用于英文环境，而且还可能用于撇号（apostrophe ），那么对单引号的任何处理都意味着额外增加了维护成本，而非减少。我觉得这里没有必要面面俱到，毕竟我们也不是出版商。
> 
> 如果这边认可 `“……”` 是不好的，那么 `‘……’` 也应被认为是不好的，所以如果要处理 `“……”` 那也要处理 `‘……’`。至于维护成本上我觉得没有什么区别，这两种情况的代码是基本一致的（如果不处理 apostrophe 那 `inner_replace` 就基本是把相同的代码重复了两遍而已）。

逻辑上两个的确是一致的，但是实践上不存在中文使用单引号的情形。替换单引号反而会干扰英文中单引号和撇号的使用。所以，替换双引号的收益是正的，替换单引号的收益是负的。我所说的维护成本，是维护文档的成本，而非维护代码的成本。代码上固然一致，但是文档中却总是要强行 escape。如果一个功能在 docs 中的所有实例都需要 escape，这个功能就没有引入的必要。

**Comment by Tiphereth-A** [comment]
> > > 如果 wiki 里没有有中文用单引号的例子，单引号只用于英文环境，而且还可能用于撇号（apostrophe ），那么对单引号的任何处理都意味着额外增加了维护成本，而非减少。我觉得这里没有必要面面俱到，毕竟我们也不是出版商。
> > 
> > 如果这边认可 `“……”` 是不好的，那么 `‘……’` 也应被认为是不好的，所以如果要处理 `“……”` 那也要处理 `‘……’`。至于维护成本上我觉得没有什么区别，这两种情况的代码是基本一致的（如果不处理 apostrophe 那 `inner_replace` 就基本是把相同的代码重复了两遍而已）。
> 
> 逻辑上两个的确是一致的，但是实践上不存在中文使用单引号的情形。替换单引号反而会干扰英文中单引号和撇号的使用。所以，替换双引号的收益是正的，替换单引号的收益是负的。我所说的维护成本，是维护文档的成本，而非维护代码的成本。代码上固然一致，但是文档中却总是要强行 escape。如果一个功能在 docs 中的所有实例都需要 escape，这个功能就没有引入的必要。

也许你可以看看最新的几个 commit，现在代码能正确处理 apostrophe，文档那边不需要再 skip 了 🤔

**Comment by c-forrest** [comment]
> > > > 如果 wiki 里没有有中文用单引号的例子，单引号只用于英文环境，而且还可能用于撇号（apostrophe ），那么对单引号的任何处理都意味着额外增加了维护成本，而非减少。我觉得这里没有必要面面俱到，毕竟我们也不是出版商。
> > > 
> > > 
> > > 如果这边认可 `“……”` 是不好的，那么 `‘……’` 也应被认为是不好的，所以如果要处理 `“……”` 那也要处理 `‘……’`。至于维护成本上我觉得没有什么区别，这两种情况的代码是基本一致的（如果不处理 apostrophe 那 `inner_replace` 就基本是把相同的代码重复了两遍而已）。
> > 
> > 
> > 逻辑上两个的确是一致的，但是实践上不存在中文使用单引号的情形。替换单引号反而会干扰英文中单引号和撇号的使用。所以，替换双引号的收益是正的，替换单引号的收益是负的。我所说的维护成本，是维护文档的成本，而非维护代码的成本。代码上固然一致，但是文档中却总是要强行 escape。如果一个功能在 docs 中的所有实例都需要 escape，这个功能就没有引入的必要。
> 
> 也许你可以看看最新的几个 commit，现在代码能正确处理 apostrophe，文档那边不需要再 skip 了 🤔

我就是说的最新的 commit 里边，所有英文单引号都被 skip 了（

**Comment by Tiphereth-A** [comment]
> 我就是说的最新的 commit 里边，所有英文单引号都被 skip 了（

那个是 expected behavior，GCC 的输出就是这个格式

**Comment by c-forrest** [comment]
> > 我就是说的最新的 commit 里边，所有英文单引号都被 skip 了（
> 
> 那个是 expected behavior，GCC 的输出就是这个格式

问题就是 wiki 里只有这种单引号。

现在我们考虑如下提案：将 wiki 中所有 ‘……’ 都替换为 『……』。我们不要先入为主地将它理解为将中文中的单引号替换为全角引号。我们就看 wiki 现有的用例。结果发现全是英文单引号，而且还是 GCC 输出结果，连改的余地都没有。那么，我们这个提案在 wiki 的语境中就应该理解为：将英文中的单引号全部替换为全角引号。这种提案显然是没有意义的。因为目前的替换代码中没有做到替换中文的单引号，那么预期行为就是将所有单引号均替换，而所有单引号在 wiki 里就只有英文单引号，那么这一系列行为的实际效果就是只替换了英文单引号。但是我们又不希望这样做，只能 skip。所以一通操作，唯一的结果就是将所有相关段落都加上了 skip。所以，我的看法是，我们一开始就不要引入替换单引号的部分，它没有起到应有的作用，我们也不预期它起到任何作用（因为我们没有中文单引号的用例），但是却会起到一定的反作用。

image

**Comment by Tiphereth-A** [comment]
> > 我就是说的最新的 commit 里边，所有英文单引号都被 skip 了（
> 
> 那个是 expected behavior，GCC 的输出就是这个格式

修正一下这句话，我复现失败了，这个引号应该本来就是 U+0027，可能是受 smart quotes 特性之类的影响才变成了 U+2018 U+2019。

GCC 那边似乎也是以 U+0027 为主，see <https://github.com/gcc-mirror/gcc/blob/33b945b4e637f4e46e0c3bc42bded2949124c940/gcc/testsuite/gdc.dg/pr97002.d#L38-L39>

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/gcc-mirror/gcc/blob/33b945b4e637f4e46e0c3bc42bded2949124c940/gcc/testsuite/gdc.dg/pr97002.d">gcc/gcc/testsuite/gdc.dg/pr97002.d at 33b945b4e637f4e46e0c3bc42bded2949124c940 · gcc-mirror/gcc</a></strong></div><div>Contribute to gcc-mirror/gcc development by creating an account on GitHub.</div></blockquote>

File: scripts/linter/postprocess.py [hunk 2]
```diff
@@ -33,10 +34,102 @@ def fix_full_stop(md_content: str, **kwargs):
...
+                case '‘':
+                    if stack and stack[-1] == '‘':
+                        log(f"extra left single quote detected, treating as right quote",
+                            type=f'warning line={ln + 1},col={col + 1}')
+                        return inner_replace(ln, col, '’')
+                    stack.append('‘')
+                    return '『'
+                case '’':
+                    nonlocal line
+                    if len(line) > 1 and (line[col - 1].isascii() if col > 0 else True) and (line[col + 1].isascii() if col < len(line) - 1 else True):
```
Comment (c-forrest) [inline scripts/linter/postprocess.py]: 更精确一点应该是左侧字符 isalpha() 右侧字符 isalpha() or isspace()？比如 「请直接打印出 'Hello world!' 并返回 0。」按照目前的逻辑是不是就会判断为撇号而不是单引号？

Comment (Tiphereth-A) [reply scripts/linter/postprocess.py]: isalpha 肯定是错的。烫知识，`'汉字'.isalpha() == True`

Comment (Tiphereth-A) [reply scripts/linter/postprocess.py]: 想了下判后一个字符是没必要的，现在就先只判前一个字符是不是字母或数字了

Comment (c-forrest) [reply scripts/linter/postprocess.py]: emmm，所以现在改了之后，「请直接打印出 ‘Hello world’ 并返回 0。」会被怎么处理？

Comment (Tiphereth-A) [reply scripts/linter/postprocess.py]: https://github.com/Tiphereth-A/OI-wiki/blob/80d11a9a2e6fb22e558981edbb2df595bef47f63/test/linter/test_postprocess.py#L439

btw，现在会忽略输入里引号的左右，实际的左右用栈维护

Comment (c-forrest) [reply scripts/linter/postprocess.py]: 其实我这个好像是再下面那个 case。不过无所谓了，反正都是一些 corner case，几乎碰不到这些。
